contracts/BSCValidatorSet.sol

```diff
@@ -479,7 +546,190 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
+          numOfFelony++;
+          break;
+        }
+      }
+    }
+
+    // 2. clear all maintain info
+    delete maintainingValidatorSet;
+    for (uint256 i = 0; i < validatorSet.length; i++) {
+      delete maintainInfoMap[validatorSet[i].consensusAddress];
+    }
+
+    // 3. get unjailed validators from validatorSet
+    unjailedValidatorSet = new Validator[](validatorSet.length - numOfFelony);
+    uint256 i = 0;
+    for (uint index = 0; index < validatorSet.length; index++) {
+      if (!validatorSet[index].jailed) {
+        unjailedValidatorSet[i] = validatorSet[index];
+        i++;
+      }
+    }
+
+    return unjailedValidatorSet;
+  }
+
+  function _enterMaintenance(address validator, uint256 index) private {
+    // step 1: modify status of the validator
+    MaintainInfo storage maintainInfo = maintainInfoMap[validator];
+    maintainInfo.isMaintaining = true;
+    maintainInfo.startBlockNumber = block.number;
+
+    // step 2: add the validator to maintainingValidatorSet
+    maintainingValidatorSet.push(currentValidatorSet[index]);
+    maintainInfo.index = maintainingValidatorSet.length;
+
+    // step 3: remove the validator from currentValidatorSet
+    _removeFromCurrentValidatorSet(validator, index);
+
+    emit validatorEnterMaintenance(validator);
+  }
+
+  function _exitMaintenance(address validator) private returns (bool isFelony){
+    // step 1: modify status of the validator
+    MaintainInfo storage maintainInfo = maintainInfoMap[validator];
+    maintainInfo.isMaintaining = false;
+
+    // step 2: add the validator to currentValidatorSet
+    // the actual index
+    uint256 index = maintainInfo.index - 1;
+    currentValidatorSet.push(maintainingValidatorSet[index]);
+    currentValidatorSetMap[validator] = currentValidatorSet.length;
+
+    // step 3: remove the validator from maintainingValidatorSet
+    // 0 is the `sentinel value`
+    maintainInfo.index = 0;
+    // It is ok that the maintainingValidatorSet is not in order.
+    if (index != maintainingValidatorSet.length - 1) {
+      maintainingValidatorSet[index] = maintainingValidatorSet[maintainingValidatorSet.length - 1];
+      maintainInfoMap[maintainingValidatorSet[index].consensusAddress].index = index + 1;
+    }
+    maintainingValidatorSet.pop();
+    emit validatorExitMaintenance(validator);
+
+    // step 4: slash
+    if (maintainSlashScale == 0 || currentValidatorSet.length == 0) {
+      // should not happen, still protect
+      return false;
+    }
+
+    uint256 slashCount =
+      block.number.sub(maintainInfo.startBlockNumber).div(currentValidatorSet.length).div(maintainSlashScale);
+
+    (uint256 misdemeanorThreshold, uint256 felonyThreshold) = ISlashIndicator(SLASH_CONTRACT_ADDR).getSlashThresholds();
+    isFelony = false;
+    if (slashCount >= felonyThreshold) {
+      _felony(validator);
+      ICrossChain(CROSS_CHAIN_CONTRACT_ADDR).sendSynPackage(SLASH_CHANNELID, encodeSlashPackage(validator), 0);
```

Comment:
Validator contract do not have the permission to send `SLASH_CHANNELID` package, now only slash contract is allowed to do so, you may check the `crosschain` contract.

Reply:
fixed